File: LCObjective-C/LCObjective-C/Problems/MergeTwoSortedLists/MergeTwoSortedLists.m
//
//  MergeTwoSortedLists.m
//  LCObjective-C
//
//  Created by Sherzod Akhmedov on 18/07/22.
//

#import <Foundation/Foundation.h>
#import "MergeTwoSortedLists.h"
#import "ListNode.h"

@interface MergeTwoSortedLists ()

+ (NSArray *)convertListNodeToArray: (ListNode * _Nullable)node;

@end


@implementation MergeTwoSortedLists

+ (ListNode *)mergeTwoLists:(ListNode *)list1 andList:(ListNode *)list2 {
    NSArray *nums = [[self convertListNodeToArray:list1] arrayByAddingObjectsFromArray:[self convertListNodeToArray:list2]];
    
    nums = [nums sortedArrayUsingSelector:@selector(compare:)];
    if ([nums count] == 0) return nil;
    ListNode *node = [[ListNode alloc] initWithVal:[[nums objectAtIndex:0] longValue]];
    ListNode *nextListNode = node;
    
    if ([nums count] > 1) {
        for (NSUInteger i = 1; i < [nums count]; i++) {
            NSInteger num = [[nums objectAtIndex:i] longValue];
            [nextListNode setNext:[[ListNode alloc] initWithVal:num]];
            nextListNode = [nextListNode next];
        }
    }
    
    return node;
}

+ (NSArray *)convertListNodeToArray:(ListNode *)node {
    if (node == nil) return [[NSArray alloc] init];
    NSMutableArray *result = [[NSMutableArray alloc] init];
    ListNode *sequenceNode = node;
    for (;;) {
        NSNumber *val = [[NSNumber alloc] initWithLong:[sequenceNode val]];
        [result insertObject:val atIndex:[result count]];
        
        if ([sequenceNode next] == nil) {
            break;
        }
        sequenceNode = [sequenceNode next];
    }
    
    return result;
}

@end

File: LCSwift/LCSwift/Problems/MergeTwoSortedLists/MergeTwoSortedLists.swift
//
//  MergeTwoSortedLists.swift
//  LCSwift
//
//  Created by Sherzod Akhmedov on 18/07/22.
//

import Foundation

class MergeTwoSortedLists {
    
    class ListNode {
        public var val: Int
        public var next: ListNode?
        public init() { self.val = 0; self.next = nil; }
        public init(_ val: Int) { self.val = val; self.next = nil; }
        public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
    }
    
    static func mergeTwoLists(_ list1: ListNode?, _ list2: ListNode?) -> ListNode? {
        let nums: [Int] = (convertListNodeToArray(list1) + convertListNodeToArray(list2)).sorted(by: { $0 < $1 })
        guard !nums.isEmpty else { return nil }
        
        let node = ListNode(nums.first!)
        var nextListNode: ListNode = node
        for num in nums.dropFirst() {
            nextListNode.next = ListNode(num)
            nextListNode = nextListNode.next!
        }
        
        return node
    }
    
    private static func convertListNodeToArray(_ list: ListNode?) -> [Int] {
        guard let list = list else { return [] }
        var result: [Int] = []
        var sequenceNode: ListNode = list
        while true {
            result.append(sequenceNode.val)
            guard sequenceNode.next != nil else { break }
            sequenceNode = sequenceNode.next!
        }
        return result
    }
}
